### PR TITLE
Switches to positive wording for all rules besides 2.3,

### PR DESCRIPTION
**Pull request by viktorklang**
simply because I could not find appropriate wording.

**NOTE: No intentional semantical changes have been made. If you find issues, please, please, please let me know**

Harmonises the use of "call" and "invoke" to use "call" in all places
where "invoke" was used.

@reactive-streams/contributors Could you please sanity check the changes proposed here? Thanks!
Please vote/comment before the 21st of January (as always, if you need more time let us know)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -134,11 +134,11 @@ public interface Subscription {
 
 | ID                        | Rule                                                                                                   |
 | ------------------------- | ------------------------------------------------------------------------------------------------------ |
-| <a name="3.1">1</a>       | `Subscription.request` or `Subscription.cancel` MUST not be called outside of its `Subscriber` context. A `Subscription` represents the unique relationship between a `Subscriber` and a `Publisher` [see [2.12](#2.12)] |
+| <a name="3.1">1</a>       | `Subscription.request` and `Subscription.cancel` MUST only be called inside of its `Subscriber` context. A `Subscription` represents the unique relationship between a `Subscriber` and a `Publisher` [see [2.12](#2.12)] |
```
Comment (viktorklang) [inline README.md]: Switched from `or` to `and` here, since that was the intent. Is it clearer?

File: README.md [hunk 2]
```diff
@@ -78,7 +78,7 @@ public interface Publisher<T> {
 
 | ID                        | Rule                                                                                                   |
 | ------------------------- | ------------------------------------------------------------------------------------------------------ |
-| <a name="1.1">1</a>       | The number of `onNext` signaled by a `Publisher` to a `Subscriber` MUST NOT exceed the cumulative demand that has been signaled via that `Subscriber`’s `Subscription` |
+| <a name="1.1">1</a>       | The number of `onNext` signals sent by a `Publisher` to a `Subscriber` MUST always be less than or equal to the elements requested by that `Subscriber`´s `Subscription` |
```
Comment (viktorklang) [inline README.md]: Is `signals sent` better than `signaled` here?

File: README.md [hunk 3]
```diff
@@ -134,11 +134,11 @@ public interface Subscription {
...
+| <a name="3.1">1</a>       | `Subscription.request` and `Subscription.cancel` MUST only be called inside of its `Subscriber` context. A `Subscription` represents the unique relationship between a `Subscriber` and a `Publisher` [see [2.12](#2.12)] |
 | <a name="3.2">2</a>       | The `Subscription` MUST allow the `Subscriber` to call `Subscription.request` synchronously from within `onNext` or `onSubscribe` |
-| <a name="3.3">3</a>       | `Subscription.request` MUST NOT allow unbounded recursion such as `Subscriber.onNext` -> `Subscription.request` -> `Subscriber.onNext` |
-| <a name="3.4">4</a>       | `Subscription.request` SHOULD NOT synchronously perform heavy computations that would impact its callers responsivity |
-| <a name="3.5">5</a>       | `Subscription.cancel` MUST NOT synchronously perform heavy computations, MUST be idempotent and MUST be thread-safe |
+| <a name="3.3">3</a>       | `Subscription.request` MUST only allow bounded recursion such as `Subscriber.onNext` -> `Subscription.request` -> `Subscriber.onNext` |
```
Comment (viktorklang) [inline README.md]: Is this semantically equivalent to the old version?

File: README.md [hunk 4]
```diff
@@ -134,11 +134,11 @@ public interface Subscription {
...
+| <a name="3.1">1</a>       | `Subscription.request` and `Subscription.cancel` MUST only be called inside of its `Subscriber` context. A `Subscription` represents the unique relationship between a `Subscriber` and a `Publisher` [see [2.12](#2.12)] |
 | <a name="3.2">2</a>       | The `Subscription` MUST allow the `Subscriber` to call `Subscription.request` synchronously from within `onNext` or `onSubscribe` |
-| <a name="3.3">3</a>       | `Subscription.request` MUST NOT allow unbounded recursion such as `Subscriber.onNext` -> `Subscription.request` -> `Subscriber.onNext` |
-| <a name="3.4">4</a>       | `Subscription.request` SHOULD NOT synchronously perform heavy computations that would impact its callers responsivity |
-| <a name="3.5">5</a>       | `Subscription.cancel` MUST NOT synchronously perform heavy computations, MUST be idempotent and MUST be thread-safe |
+| <a name="3.3">3</a>       | `Subscription.request` MUST only allow bounded recursion such as `Subscriber.onNext` -> `Subscription.request` -> `Subscriber.onNext` |
+| <a name="3.4">4</a>       | `Subscription.request` SHOULD respect the responsivity of its caller by returning in a timely manner |
```
Comment (viktorklang) [inline README.md]: Reason for the change here is to put the focus on the responsivity not on how "heavy" a computation may or may not be.

File: README.md (hunk 2 above)
Comment (rkuhn) [inline README.md]: Yes, that’s better, we can improve it further by replacing “always” with “at all times” to put focus on the continued need to abide by this constraint.

File: README.md [hunk 5]
```diff
@@ -118,8 +118,8 @@ public interface Subscriber<T> {
...
 | <a name="2.10">10</a>     | A `Subscriber` MUST be prepared to receive an `onError` signal with or without a preceding `Subscription.request(long n)` call |
 | <a name="2.11">11</a>     | A `Subscriber` MUST make sure that all calls on its `onXXX` methods happen-before [[1]](#footnote-2-1) the processing of the respective signals. I.e. the Subscriber must take care of properly publishing the signal to its processing logic |
-| <a name="2.12">12</a>     | `Subscriber.onSubscribe` MUST NOT be called more than once (based on object equality) |
-| <a name="2.13">13</a>     | Invoking `onSubscribe`, `onNext`, `onError` or `onComplete` MUST return normally. The only legal way for a `Subscriber` to signal failure is by cancelling its `Subscription`. In the case that this rule is violated, any associated `Subscription` to the `Subscriber` MUST be considered as cancelled, and the invoker MUST raise this error condition in a fashion that is adequate for the runtime environment |
+| <a name="2.12">12</a>     | `Subscriber.onSubscribe` MUST be called at most once (based on object equality) |
```
Comment (rkuhn) [inline README.md]: Even clearer: “… at most once _for a given Subscriber_ …”

File: README.md (hunk 1 above)
Comment (rkuhn) [inline README.md]: yes, this is better

File: README.md (hunk 3 above)
Comment (rkuhn) [inline README.md]: Since there are many different other kinds of behaviors not related to recursion, I would suggest “Subscription.request MUST place an upper bound on possible recursion between Publisher and Subscriber (an example for undesirable open recursion would be `onNext -> request -> onNext -> ...`).”

File: README.md (hunk 4 above)
Comment (rkuhn) [inline README.md]: Should we keep “performing heavy computations” as an undesirable example in parentheses?

**Comment by rkuhn** [comment]
Very good initiative, avoiding negation makes the intent clearer in all those cases.

File: README.md (hunk 2 above)
Comment (viktorklang) [inline README.md]: Great suggestion, thanks

Comment (drewhk) [inline README.md]: Now looking at this rule, there is a slight imprecision here. "less than or equal to the elements requested" is a bit vague and can be interpreted as the last request(X) call. What about:

"The total number of `onNext` signals sent by a `Publisher` to a `Subscriber` until any point in time MUST always be less than or equal to the total elements requested by that `Subscriber`´s `Subscription until that time."

Maybe reworded, my point is to make it clear that "elements requested" is meant as an aggregate, not just the last request() call.

File: README.md (hunk 5 above)
Comment (viktorklang) [inline README.md]: That's clearer!

**Comment by drewhk** [comment]
LGTM

File: README.md (hunk 3 above)
Comment (viktorklang) [inline README.md]: Good feedback, I'm going to update this section with a clearer version with a footnote explaining the reason.

File: README.md (hunk 4 above)
Comment (viktorklang) [inline README.md]: Footnote?

Comment (viktorklang) [inline README.md]: Adding footnote

**Comment by viktorklang** [comment]
@rkuhn Thanks for the feedback, I owe most of it to @DougLea for not only raising the issue but also suggesting how to reword it.

File: README.md (hunk 2 above)
Comment (viktorklang) [inline README.md]: @drewhk Great point, I've updated the PR.

File: README.md [hunk 6]
```diff
@@ -147,12 +147,16 @@ public interface Subscription {
...
+| <a name="3.17">17</a>     | A `Subscription` MUST support an unbounded number of calls to request and MUST support a demand (sum requested - sum delivered) up to 2^63-1 (`java.lang.Long.MAX_VALUE`). A demand of exactly 2^63-1 (`java.lang.Long.MAX_VALUE`) MAY be considered by the `Publisher` as `effectively unbounded`[[3](#footnote-3-3)]. If demand becomes higher than 2^63-1 then the `Publisher` MUST signal an onError with `java.lang.IllegalStateException` on the given `Subscriber`. The cause message MUST include a reference to this rule and/or quote the full rule |
 
-[<a name="footnote-3-1">1</a>] : As it is not feasibly reachable with current or forseen hardware within a reasonable amount of time (1 element per nanosecond would take 292 years) to fulfill a demand of 2^63-1.
+[<a name="footnote-3-1">1</a>] : An example for undesirable synchronous, open recursion would be `Subscriber.onNext` -> `Subscription.request` -> `Subscriber.onNext` -> …, as it very quickly would result in blowing the calling Thread´s stack.
+
+[<a name="footnote-3-2">2</a>] : Avoid heavy computations and other things that would stall the caller´s thread of execution
```
Comment (viktorklang) [inline README.md]: @rkuhn This good?

File: README.md [hunk 7]
```diff
@@ -147,12 +147,16 @@ public interface Subscription {
...
+| <a name="3.17">17</a>     | A `Subscription` MUST support an unbounded number of calls to request and MUST support a demand (sum requested - sum delivered) up to 2^63-1 (`java.lang.Long.MAX_VALUE`). A demand of exactly 2^63-1 (`java.lang.Long.MAX_VALUE`) MAY be considered by the `Publisher` as `effectively unbounded`[[3](#footnote-3-3)]. If demand becomes higher than 2^63-1 then the `Publisher` MUST signal an onError with `java.lang.IllegalStateException` on the given `Subscriber`. The cause message MUST include a reference to this rule and/or quote the full rule |
 
-[<a name="footnote-3-1">1</a>] : As it is not feasibly reachable with current or forseen hardware within a reasonable amount of time (1 element per nanosecond would take 292 years) to fulfill a demand of 2^63-1.
+[<a name="footnote-3-1">1</a>] : An example for undesirable synchronous, open recursion would be `Subscriber.onNext` -> `Subscription.request` -> `Subscriber.onNext` -> …, as it very quickly would result in blowing the calling Thread´s stack.
```
Comment (viktorklang) [inline README.md]: @rkuhn this good?

**Comment by viktorklang** [comment]
@reactive-streams/contributors Merging!
